content/blog/2020-11-09-modern-gc/index.md

```diff
@@ -0,0 +1,80 @@
+
+## Evaluation
+
+The evaluation for this paper is rather extensive and follows many of the guidelines we would hope for such as:
+
+- describing the choice of benchmarks and why some were left out
+- specifying the java virtual machine used
+- parameters like block and line size
+- warmup periods
+- a description of the hardware and software platforms being used
+- reporting both averages and confidence intervals
+
+The goal of their analysis is to show that conservative garbage collectors are not as limited as they might first appear and that a conservative garbage collector can be competitive against the state of the art jvm garbage collector.
+
+### Impact of Conservatism
+
+Probably the most obvious concern with conservative garbage collection is that dead objects will survive in the heap due to an ambiguous reference that is actually a value. However in practice, the paper found that these zombie references are not much of a concern. They measured the size of memory retained by undead objects over the memory size of live objects and found the average to be 0.02%. This would suggest that in most cases, the amount of extra objects maintained due to an ambiguous reference is miniscule.
+
+Similarly, since every ambiguous reference must be filtered against the object map, there may be a concern with the amount of overhead added to be conservative. Comparing exact Mark-Sweep with conservative Mark-Sweep, the paper found a 19.8% increase in overhead when garbage collecting. However, the effect on total runtime of the program was 3.7% which suggests that the overhead for just having to filter ambiguous pointers isn't huge.
+
+### Performance
+
+Figure 4 is probably the most important figure in the paper and sums up the performance analysis for conservative garbage collection.
+
+![Figure 4](Figure4.png)
+
+Across the six exact garbage collectors with their conservative counterparts, all but one kept runtime overhead below 3% in a straight up comparison. The authors claim that MCC is an outlier here because of the it's poor pinning granularity as shown in the previous section. If that wasn't convincing enough, when compared against Gen Immix, the state of the art gc for Jikes RVM, the conservative Immix variations all performed within a few percent excess overhead. In the paper, each benchmark is broken down in detail to show that in all cases the conservative RC Immix stays below 5% overhead of Gen Immix. The paper also verifies that the RC Immix conservative garbage collector stays competitive as available heap size increases
+
+## Discussion Questions
+
+  1. When might you choose conservative garbage collection over exact garbage collection?
+  2. Are there other ways/metrics that garbage collectors should be analyzed for efficiency? Is the analysis in this paper conclusive?
```

Comment:
A little late for this, but this could be a bit more specific—namely, it could remind the reader what the ones used here were so we can judge whether that's everything.